apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,868 @@
+/**
+ * BluetoothConnectionManager:
+ * BluetoothConnectionManager only update state and does not involve in any UI
+ *   logic.
+ *
+ * @module BluetoothConnectionManager
+ */
+define(function(require) {
+  'use strict';
+
+  var AdapterManager = require('modules/bluetooth/bluetooth_adapter_manager');
+  var AsyncStorage = require('shared/async_storage');
+  require('shared/async_storage');
+
+  var _debug = false;
+  var Debug = function() {};
+  if (_debug) {
+    Debug = function btam_debug(msg) {
+      console.log('--> [BluetoothConnectionManager]: ' + msg);
+    };
+  }
+
+  /**
+   * @alias module:modules/bluetooth/BluetoothConnectionManager
+   * @requires module:modules/bluetooth/bluetooth_adapter_manager
+   * @requires module:shared/async_storage
+   * @return {BluetoothConnectionManager}
+   */
+  var BluetoothConnectionManager = {
+    /**
+     * The profiles of connected device that we are defined here.
+     *
+     * @public
+     * @memberOf BluetoothConnectionManager
+     * @type {Object}
+     */
+    Profiles: {
+      'hfp': 0x111E,
+      'a2dp': 0x110D
+    },
+
+    /**
+     * The address of device that we are trying to connect.
+     *
+     * @public
+     * @memberOf BluetoothConnectionManager
+     * @type {String}
+     */
+    connectingAddress: null,
+
+    /**
+     * A object that we cache the connected devices information(address, device,
+     * connectedProfiles). It will be inited while default adapter is ready.
+     * And these information is coming from profile events.
+     * Each connected device is hashed by device address.
+     *
+     * EX:
+     * _connectedDevicesInfo = {
+     *   'AA:BB:CC:00:11:22': {
+     *     'device': DeviceObject,
+     *     'connectedProfiles': {
+     *       'hfp': true,
+     *       'a2dp': false
+     *     }
+     *   }
+     * };
+     *
+     * @private
+     * @memberOf BluetoothConnectionManager
+     * @type {Object}
+     */
+    _connectedDevicesInfo: {},
+
+    /**
+     * Return the cache of connected devices in ConnectionManager.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @return {Object}
+     */
+    getConnectedDevices: function btcm_getConnectedDevices() {
+      return this._connectedDevicesInfo;
```

Comment:
This function should return a promise. This differs from the case of getting an adapter because in this case gaia actively requests the latest connection status from gecko.

Consider this case, bt context calls to this function when `_initConnectedDevicesInfo` does not finish the initialization. Bt context does not get the connected device correctly until it calls to this function again.

To solve the issue, we should cache the promise returned from `_initConnectedDevicesInfo` and always wait for it resolved before returning the result.
And even in this module, we should not use `this._connectedDeviceInfo` directly but always use this function. This guarantees that the update to `this._connectedDevicesInfo` is always made after `this._connectedDevicesInfo` is initialized.